base/subarray.jl

```diff
@@ -1,60 +1,73 @@
-struct SubArray{T,N,P,I,L} <: AbstractArray{T,N}
+# The view itself supports fast linear indexing if length(indices) == 1 and IndexStyle(indices[1]) is LinearFast()
+"""
+    SubArray{T,N,P,I} <: AbstractArray{T,N}
+
+A specialized array type for that represents a view into an indexed subsection of a "parent" array.
+
+Typically constructed with [`view`](@ref) or indexing inside a [`@view`](@ref) or [`@views`](@ref) macro,
+SubArrays simply hold onto the parent array and the indices. The additional type parameters concretely
+identify the type of the parent array (`P`) and the type of the tuple of indices (`I`).
+"""
+struct SubArray{T,N,P,I} <: AbstractArray{T,N}
     parent::P
-    indexes::I
-    offset1::Int       # for linear indexing and pointer, only valid when L==true
-    stride1::Int       # used only for linear indexing
-    function SubArray{T,N,P,I,L}(parent, indexes, offset1, stride1) where {T,N,P,I,L}
-        @_inline_meta
-        check_parent_index_match(parent, indexes)
-        new(parent, indexes, offset1, stride1)
-    end
+    indices::I
 end
 # Compute the linear indexability of the indices, and combine it with the linear indexing of the parent
-function SubArray(parent::AbstractArray, indexes::Tuple)
+function SubArray(parent::AbstractArray, indices::Tuple)
     @_inline_meta
-    SubArray(IndexStyle(viewindexing(indexes), IndexStyle(parent)), parent, ensure_indexable(indexes), index_dimsum(indexes...))
+    SubArray(IndexStyle(viewindexing(indices), IndexStyle(parent)), parent, ensure_indexable(indices), index_shape(indices...))
 end
-function SubArray(::IndexCartesian, parent::P, indexes::I, ::NTuple{N,Any}) where {P,I,N}
+function SubArray(::IndexCartesian, parent::P, indices::I, ::NTuple{N,Any}) where {P,I,N}
     @_inline_meta
-    SubArray{eltype(P), N, P, I, false}(parent, indexes, 0, 0)
+    SubArray{eltype(P), N, P, I}(parent, indices)
 end
-function SubArray(::IndexLinear, parent::P, indexes::I, ::NTuple{N,Any}) where {P,I,N}
+function SubArray(::IndexLinear, parent::P, indices::I, shape::NTuple{N,Any}) where {P,I,N}
     @_inline_meta
-    # Compute the stride and offset
-    stride1 = compute_stride1(parent, indexes)
-    SubArray{eltype(P), N, P, I, true}(parent, indexes, compute_offset1(parent, stride1, indexes), stride1)
+    t = linearize_indices(parent, indices, shape)
+    SubArray{eltype(P), N, P, typeof(t)}(parent, t)
 end
 
-check_parent_index_match(parent, indexes) = check_parent_index_match(parent, index_ndims(indexes...))
-check_parent_index_match(parent::AbstractArray{T,N}, ::NTuple{N, Bool}) where {T,N} = nothing
-check_parent_index_match(parent, ::NTuple{N, Bool}) where {N} =
-    throw(ArgumentError("number of indices ($N) must match the parent dimensionality ($(ndims(parent)))"))
+# StridedSubArrays depend upon Ranges wrapped in ReshapedArrays, so more functionality
+# is defined within reshapedarray.jl
+
+# Collapse a set of linearly compatible indices into a single index
```

Comment:
Some brief docs/comments might be helpful: extracting what's going on by looking at the code is less straightforward than the idea itself.

Presumably the input here must be de-colonified? I've lost track of which function is in charge of that.

Reply:
Yes, I'll definitely need to revise the docs/devdocs.

Colon conversion is now done within `to_indices`.